util/promlint/promlint.go

```diff
@@ -0,0 +1,182 @@
+	Metric string
+
+	// A description of the issue for this Problem.
+	Text string
+}
+
+// New creates a new Linter that reads an input stream of Prometheus metrics.
+// Only the text exposition format is supported.
+func New(r io.Reader) *Linter {
+	return &Linter{
+		r: r,
+	}
+}
+
+// Lint performs a linting pass, returning a slice of Problems indicating any
+// issues found in the metrics stream.
+func (l *Linter) Lint() ([]Problem, error) {
+	// TODO(mdlayher): support for protobuf exposition format?
+	d := expfmt.NewDecoder(l.r, expfmt.FmtText)
+
+	var problems []Problem
+
+	var mf dto.MetricFamily
+	for {
+		if err := d.Decode(&mf); err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return nil, err
+		}
+
+		problems = append(problems, lint(mf)...)
+	}
+
+	// Ensure deterministic output.
+	sort.SliceStable(problems, func(i, j int) bool {
+		if problems[i].Metric < problems[j].Metric {
+			return true
+		}
+
+		return problems[i].Text < problems[j].Text
+	})
+
+	return problems, nil
+}
+
+// lint is the entry point for linting a single metric.
+func lint(mf dto.MetricFamily) []Problem {
+	// TODO(mdlayher): lint rules for specific metrics types.
+	return lintGeneral(mf)
+}
+
+// lintGeneral detects general issues not tied to any specific metric type.
+func lintGeneral(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	// Expect all metrics to have help text available.
+	if mf.Help == nil {
+		problems = append(problems, Problem{
+			Metric: *mf.Name,
+			Text:   "no help text",
+		})
+	}
+
+	problems = append(problems, lintMetricUnits(mf)...)
+
+	return problems
+}
+
+// lintMetricUnits detects issues with metric unit names.
+func lintMetricUnits(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	// Expect all metric names to use base units.
+	for _, u := range baseUnits {
+		for _, p := range unitPrefixes {
+			// Assume that metrics names will be of the form "foo_units".
+			unit := p + u
+			if !strings.Contains(*mf.Name, "_"+unit) {
+				continue
+			}
+
+			problems = append(problems, Problem{
+				Metric: *mf.Name,
+				Text:   fmt.Sprintf("use base unit %q instead of %q", u, unit),
+			})
+
+			// Special case: enforce the international spelling "metres" over the
+			// American spelling "meters".
+			if strings.Contains(*mf.Name, "meters") {
+				problems = append(problems, Problem{
+					Metric: *mf.Name,
+					Text:   `use international spelling "metres" instead of "meters"`,
+				})
+			}
+		}
+	}
+
+	// Special case: unit "grams" should be SI base unit "kilograms".
+	if strings.Contains(*mf.Name, "_grams") {
+		problems = append(problems, Problem{
+			Metric: *mf.Name,
+			Text:   `use base unit "kilograms" instead of "grams"`,
+		})
+	}
+
+	return problems
+}
+
+// Units and their possible prefixes recognized by this library.  More can be
+// added over time as needed.
+var (
+	baseUnits = []string{
+		"amperes",
+		"bits",
+		"bytes",
+		"candela",
+		"kelvins",
```

Comment:
That's what I though too, but I found:

> When spelled out or spoken, the unit is pluralised using the same grammatical rules as for other SI units such as the volt or ohm (e.g. "the triple point of water is exactly 273.16 kelvins"[6]). 
https://en.wikipedia.org/wiki/Kelvin

Thoughts?